lib/Doctrine/Common/Persistence/Mapping/AbstractClassMetadataFactory.php

```diff
@@ -167,7 +167,9 @@ public function getMetadataFor($className)
 
         if (isset($this->loadedMetadata[$realClassName])) {
             // We do not have the alias name in the map, include it
-            return $this->loadedMetadata[$className] = $this->loadedMetadata[$realClassName];
+            $this->setMetadataFor($realClassName, $this->loadedMetadata[$realClassName]);
```

Comment:
This particular change breaks in-memory caching (`$this->loadedMetadata`) for aliased classes, therefore breaking hot path at the top of this method for those:
```php
public function testGettingAliasesOnlyFetchesFromCacheOnce() : void
{
    $loadedReflection = new ReflectionProperty(AbstractClassMetadataFactory::class, 'loadedMetadata');
    $loadedReflection->setAccessible(true);

    $this->cmf->getMetadataFor(RootEntity::class);
    self::assertArrayHasKey(RootEntity::class, $loadedReflection->getValue($this->cmf));
    self::assertArrayNotHasKey('Alias:RootEntity', $loadedReflection->getValue($this->cmf));

    $this->cmf->getMetadataFor('Alias:RootEntity');
    self::assertArrayHasKey(RootEntity::class, $loadedReflection->getValue($this->cmf));
    self::assertArrayHasKey('Alias:RootEntity', $loadedReflection->getValue($this->cmf));
}
```
It's because this condition has changed semantics and no longer does what the comment above says: "_We do not have the alias name in the map, include it_"

Reply:
@Majkl578 When it is needed to change metadata for some entity in EventListeners, we do not know about aliases that were used.
When I call setMetadataFor with real class name, AbstractClassMetadataFactory should replace metadata for all aliases in $loadedMetadata. And vise versa when I call setMetadataFor with alias.
For my opinion, it is normal when we will save only real classes in $loadedMetadata, because it prevent to have many problems with aliases.
On every call getMetadataFor with alias this code will be executed
```php
// Check for namespace alias
        if (strpos($className, ':') !== false) {
            [$namespaceAlias, $simpleClassName] = explode(':', $className, 2);

            $realClassName = $this->getFqcnFromAlias($namespaceAlias, $simpleClassName);
        } else {
            $realClassName = $this->getRealClass($className);
        }
```
In order to prevent it I can create map for 'alias' => 'realClassName' and replace this code
```php
        // Check for namespace alias
        if (strpos($className, ':') !== false) {
            [$namespaceAlias, $simpleClassName] = explode(':', $className, 2);

            $realClassName = $this->getFqcnFromAlias($namespaceAlias, $simpleClassName);
        } else {
            $realClassName = $this->getRealClass($className);
        }
```
with something like this
```php
        switch (true) {
            case isset($this->aliasesMap[$className]):
                $realClassName = $this->aliasesMap[$className];
                break;
            case strpos($className, ':') !== false // Check for namespace alias
                [$namespaceAlias, $simpleClassName] = explode(':', $className, 2);
                $realClassName = $this->getFqcnFromAlias($namespaceAlias, $simpleClassName);
                break;
            default:
                $realClassName = $this->getRealClass($className);
        }
```
In this case AbstractClassMetadataFactory will not be too coupled with aliases in setMetadataFor/getMetadataFor methods.

Reply:
So, fixed.
For now in in-memory cache (also in MetadataCache) will be saved single value for all aliases. 
Calling setMetadataFor and getMetadataFor with aliases or FQCN will same keys.